src/planner/plan_util.cpp

```diff
@@ -96,5 +100,148 @@ const std::set<oid_t> PlanUtil::GetAffectedIndexes(
   return (index_oids);
 }
 
+const std::vector<col_triplet> PlanUtil::GetAffectedColumns(
+    catalog::CatalogCache &catalog_cache,
+    std::unique_ptr<parser::SQLStatementList> sql_stmt_list,
+    const std::string &db_name) {
+  std::vector<col_triplet> column_oids;
+  std::string table_name;
+  oid_t database_id, table_id;
+
+  // Assume that there is only one SQLStatement in the list
+  auto sql_stmt = sql_stmt_list->GetStatement(0);
+  switch (sql_stmt->GetType()) {
+    // For INSERT, DELETE, all columns are affected
+    case StatementType::INSERT: {
+      auto &insert_stmt = static_cast<parser::InsertStatement &>(*sql_stmt);
+      table_name = insert_stmt.GetTableName();
+    }
+      PELOTON_FALLTHROUGH;
+    case StatementType::DELETE: {
+      if (table_name.empty() || db_name.empty()) {
+        auto &delete_stmt = static_cast<parser::DeleteStatement &>(*sql_stmt);
+        table_name = delete_stmt.GetTableName();
+      }
+      auto db_object = catalog_cache.GetDatabaseObject(db_name);
+      auto table_object = db_object->GetTableObject(table_name);
+      database_id = db_object->GetDatabaseOid();
+      table_id = table_object->GetTableOid();
+      for (auto &column : table_object->GetColumnObjects()) {
+        column_oids.emplace_back(database_id, table_id, column.first);
+      }
+    } break;
+
+    // For UPDATE, columns in UpdateClause are affected
+    case StatementType::UPDATE: {
+      auto &update_stmt = static_cast<parser::UpdateStatement &>(*sql_stmt);
+      table_name = update_stmt.table->GetTableName();
+      auto db_object = catalog_cache.GetDatabaseObject(db_name);
+      auto table_object = db_object->GetTableObject(table_name);
+      database_id = db_object->GetDatabaseOid();
+      table_id = table_object->GetTableOid();
+
+      auto &update_clauses = update_stmt.updates;
+      std::set<oid_t> update_oids;
+      for (const auto &update_clause : update_clauses) {
+        LOG_TRACE("Affected column name for table(%s) in UPDATE query: %s",
+                  table_name.c_str(), update_clause->column.c_str());
+        column_oids.emplace_back(
+            database_id, table_id,
+            table_object->GetColumnObject(update_clause->column)
+                ->GetColumnId());
+      }
+    } break;
```

Comment:
We're not really looking for "affected" columns here. Probably `GetIndexableColumns` is a better name. For our purpose here, we don't need to include columns to change in INSERT/UPDATE/DELETE statements (if we ignore index update cost for now). Instead, we should include the columns in the where clause of UPDATE and DELETE statements, and return nothing for INSERT statements.